### PR TITLE
Improve handling of pills in the composer

### PR DESCRIPTION
**Pull request by SimonBrandner**
Fixes https://github.com/vector-im/element-web/issues/10134
Fixes https://github.com/vector-im/element-web/issues/10896
Fixes https://github.com/vector-im/element-web/issues/15037

<hr>

https://user-images.githubusercontent.com/25768714/125251146-4b1e2b80-e2f7-11eb-97d8-5ac4fbc76f6a.mp4

### CODE REVIEW

File: src/editor/parts.ts [hunk 1]
```diff
@@ -364,6 +373,9 @@ class RoomPillPart extends PillPart {
     get className() {
         return "mx_RoomPill mx_Pill";
     }
+
+    // FIXME: We do this to shut up the linter, is there a way to do this properly
+    onClick = undefined;
```
Comment (SimonBrandner) [inline src/editor/parts.ts:378]: Is there any way?

Comment (t3chguy) [reply src/editor/parts.ts:378]: well the issue is you say its abstract which means it must be set on all children, see suggestion above

File: src/editor/parts.ts [hunk 2]
```diff
@@ -402,6 +414,13 @@ class UserPillPart extends PillPart {
         this._setAvatarVars(node, avatarUrl, initialLetter);
     }
 
+    onClick = () => {
```
Comment (t3chguy) [inline src/editor/parts.ts:417]:
```suggestion
    protected onClick = () => {
```

File: res/css/views/rooms/_BasicMessageComposer.scss [hunk 3]
```diff
@@ -47,6 +47,7 @@ limitations under the License.
         &.mx_BasicMessageComposer_input_shouldShowPillAvatar {
             span.mx_UserPill, span.mx_RoomPill {
                 position: relative;
+                cursor: pointer;
```
Comment (t3chguy) [inline res/css/views/rooms/_BasicMessageComposer.scss:50]: I think this wants to just target User Pills and not Room Pills as those don't have an onClick function

**Review by t3chguy — DISMISSED** [review]
LGTM otherwise

File: res/css/views/rooms/_BasicMessageComposer.scss [hunk 4]
```diff
@@ -66,6 +65,14 @@ limitations under the License.
...
                 }
             }
+
+            span.mx_UserPill {
+                cursor: pointer;
+            }
+
+            span.mx_RoomPill {
+                cursor: default;
+            }
```
Comment (t3chguy) [inline res/css/views/rooms/_BasicMessageComposer.scss:75]: is this necessary? as in, is it not implied by lack of it

Comment (SimonBrandner) [reply res/css/views/rooms/_BasicMessageComposer.scss:75]: We get the text cursor thingy without it

File: src/components/views/rooms/BasicMessageComposer.tsx [hunk 5]
```diff
@@ -515,6 +516,29 @@ export default class BasicMessageEditor extends React.Component<IProps, IState>
...
     };
 
+    /**
+     * Because pills have contentEditable="false" there is no event emitted when
+     * the user tries to delete them. Therefore we need to fake what would
+     * normally happen
+     * @param direction in which to delete
+     * @returns handled
+     */
+    private fakeDeletion(direction: "deleteContentForward" | "deleteContentBackward" ): boolean {
```
Comment (t3chguy) [inline src/components/views/rooms/BasicMessageComposer.tsx:526]: nit but `direction` is a bad name for this, IMO it should be a `backward: boolean` to simplify all the things. If you want to keep the string type then maybe `action`?

Comment (t3chguy) [reply src/components/views/rooms/BasicMessageComposer.tsx:526]: and I guess the `"deleteContentBackward" : "deleteContentForward"` bit would move onto line 538

Comment (SimonBrandner) [reply src/components/views/rooms/BasicMessageComposer.tsx]: Yeah looks much nicer

Comment (t3chguy) [reply src/components/views/rooms/BasicMessageComposer.tsx]: also maybe `fakePillDeletion`

Comment (SimonBrandner) [reply src/components/views/rooms/BasicMessageComposer.tsx]: No, not really, it is also used for deleting normal text as long as it isn't a range

File: src/editor/parts.ts [hunk 6]
```diff
@@ -399,14 +408,21 @@ class UserPillPart extends PillPart {
         if (avatarUrl === defaultAvatarUrl) {
             initialLetter = Avatar.getInitialLetter(name);
         }
-        this._setAvatarVars(node, avatarUrl, initialLetter);
+        this.setAvatarVars(node, avatarUrl, initialLetter);
     }
 
-    get type(): IPillPart["type"] {
+    protected onClick = (): void => {
```
Comment (turt2live) [inline src/editor/parts.ts:414]: this should be below the getters at a minimum

**Review by turt2live — APPROVED** [review]
overall looks fine. The history of changing the composer has me worried, so will set a reminder to land this after the RC branches are cut for maximum testing.

**Review by turt2live — APPROVED** [review]

**Pull request merged**